pkg/util/util.go

```diff
@@ -1,9 +1,54 @@
 package util
 
 type StorageClassEnforcement struct {
-	AllowList    []string `yaml:"allowList"`
-	AllowAll     bool     `yaml:"allowAll"`
-	AllowDefault bool     `yaml:"allowDefault"`
+	AllowList              []string                 `yaml:"allowList"`
+	AllowAll               bool                     `yaml:"allowAll"`
+	AllowDefault           bool                     `yaml:"allowDefault"`
+	StorageSnapshotMapping []StorageSnapshotMapping `yaml:"storageSnapshotMapping,omitempty"`
+}
+
+type StorageSnapshotMapping struct {
+	VolumeSnapshotClasses []InfraToTenantMapping `yaml:"volumeSnapshotClasses,omitempty"`
+	StorageClasses        []InfraToTenantMapping `yaml:"storageClasses"`
+}
+
+type InfraToTenantMapping struct {
+	Infra  string `yaml:"infra"`
+	Tenant string `yaml:"tenant"`
+}
+
+func (sce *StorageClassEnforcement) GetInfraSnapshotClassesFromStorageClassName(storageClassName string) []string {
+	for _, storageSnapshotMapping := range sce.StorageSnapshotMapping {
+		infraStorageClasses := []string{}
+		infraSnapshotClasses := []string{}
+		for _, storageClasses := range storageSnapshotMapping.StorageClasses {
+			infraStorageClasses = append(infraStorageClasses, storageClasses.Infra)
+		}
+		for _, snapshotClasses := range storageSnapshotMapping.VolumeSnapshotClasses {
+			infraSnapshotClasses = append(infraSnapshotClasses, snapshotClasses.Infra)
+		}
+		if Contains(infraStorageClasses, storageClassName) {
+			return infraSnapshotClasses
+		}
+	}
+	return nil
+}
+
+func (sce *StorageClassEnforcement) GetTenantSnapshotClassesFromInfraStorageClassName(storageClassName string) []string {
+	for _, storageSnapshotMapping := range sce.StorageSnapshotMapping {
+		infraStorageClasses := []string{}
+		tenantSnapshotClasses := []string{}
+		for _, storageClasses := range storageSnapshotMapping.StorageClasses {
+			infraStorageClasses = append(infraStorageClasses, storageClasses.Infra)
+		}
+		for _, snapshotClasses := range storageSnapshotMapping.VolumeSnapshotClasses {
+			tenantSnapshotClasses = append(tenantSnapshotClasses, snapshotClasses.Tenant)
+		}
+		if Contains(infraStorageClasses, storageClassName) {
+			return tenantSnapshotClasses
+		}
+	}
+	return nil
```

Comment:
the controller has access to the tenant volumesnapshotclasses.

If the admin provides us a mapping with what infra SC and VSC are compatible, then we should be able to correlate that to what tenantSnapshotClasses are compatible.

Reply:
Actually this controller does NOT have access to the tenant volumesnapshot classes (as it has no kubernetes client at all) I had assumed it had this access, but access to the tenant cluster happens through the side cars (csi-provisioner, csi-snapshot) and the csi-driver container doesn't have access at all. I can potentially add a kubernetes client, but in general a CSI driver should not assume it is running in kubernetes so I would like to avoid this. In this particular case I don't think it can run outside of kubernetes since it relies on KubeVirt, but I would like to try and stick to normal csi behavior as much as possible.

Reply:
> Actually this controller does NOT have access to the tenant volumesnapshot classes (as it has no kubernetes client at all) I had assumed it had this access, but access to the tenant cluster happens through the side cars (csi-provisioner, csi-snapshot) and the csi-driver container doesn't have access at all. 

Maybe it's this simple? I know the tenant client is already being created during initialization. 

```
diff --git a/cmd/kubevirt-csi-driver/kubevirt-csi-driver.go b/cmd/kubevirt-csi-driver/kubevirt-csi-driver.go
index f7864d9a..babb85de 100644
--- a/cmd/kubevirt-csi-driver/kubevirt-csi-driver.go
+++ b/cmd/kubevirt-csi-driver/kubevirt-csi-driver.go
@@ -120,6 +120,7 @@ func handle() {
 
        driver := service.NewKubevirtCSIDriver(virtClient,
                identityClientset,
+               tenantClientSet,
                *infraClusterNamespace,
                infraClusterLabelsMap,
                storageClassEnforcement,
diff --git a/pkg/service/controller.go b/pkg/service/controller.go
index ede46c4e..e9b58d7c 100644
--- a/pkg/service/controller.go
+++ b/pkg/service/controller.go
@@ -17,6 +17,7 @@ import (
        "k8s.io/apimachinery/pkg/api/resource"
        v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
        "k8s.io/apimachinery/pkg/util/wait"
+       "k8s.io/client-go/kubernetes"
        "k8s.io/klog/v2"
        "k8s.io/utils/ptr"
        kubevirtv1 "kubevirt.io/api/core/v1"
@@ -40,6 +41,7 @@ var (
 
 // ControllerService implements the controller interface. See README for details.
 type ControllerService struct {
+       tenantClientset         *kubernetes.Clientset
        virtClient              client.Client
        infraClusterNamespace   string
        infraClusterLabels      map[string]string
diff --git a/pkg/service/driver.go b/pkg/service/driver.go
index 7fbbe15e..bd9fb338 100644
--- a/pkg/service/driver.go
+++ b/pkg/service/driver.go
@@ -25,6 +25,7 @@ type KubevirtCSIDriver struct {
 
 func NewKubevirtCSIDriver(virtClient kubevirt.Client,
        identityClientset kubernetes.Interface,
+       tenantClientset *kubernetes.Clientset,
        infraClusterNamespace string,
        infraClusterLabels map[string]string,
        storageClassEnforcement util.StorageClassEnforcement,
@@ -37,6 +38,7 @@ func NewKubevirtCSIDriver(virtClient kubevirt.Client,
 
        if runControllerService {
                d.ControllerService = &ControllerService{
+                       tenantClientset:         tenantClientset,
                        virtClient:              virtClient,
                        infraClusterNamespace:   infraClusterNamespace,
                        infraClusterLabels:      infraClusterLabels,

```